Common/app/mqtt/mqtt_agent_task.c

```diff
@@ -1688,6 +1688,29 @@ MQTTStatus_t MqttAgent_UnSubscribeSync( MQTTAgentHandle_t xHandle,
                 }
             }
 
+            ( void ) xUnlockSubCtx( pxCtx );
+        }
+        else
+        {
+            xStatus = MQTTIllegalState;
+            LogError( "Failed to acquire MQTTAgent mutex." );
+        }
+
+        /* Send unsubscribe request if only one callbacks left for this subscription */
+        if( ( xStatus == MQTTSuccess ) &&
+            ( pxCtx->pulSubCbCount[ uxSubInfoIdx ] == 1 ) )
```

Comment:
I think there is an issue with the following being called from within an unlocked state:

 if( ( xStatus == MQTTSuccess ) &&
            ( pxCtx->pulSubCbCount[ uxSubInfoIdx ] == 1 ) )
            ...
            
I think somehow the pxCtx->[uxSubInfoIdx] must be marked invalid within the locked state before calling prvSendUnsubRequest from an unlocked state and having prvSendUnsubRequest reset the record entries when it is complete. prvSendUnsubRequest being the only owner after the marking of that entry as invalid form within the locked state.

The other thing to watch out for is that the Compress functions which reorganize that array. The one indexed by [ uxSubInfoIdx ] cannot be called with any entries in an IN_PROGRESS state. For that reason I would mark it PENDING (not moveable), copy the necessary info to perform the unsubscribe, and then reset it as AVAIL. The Compress functions would have to block when any entry is pending. I would jut have a top-level pending mutex to protect against that.

Reply:
I believe the concurrency issues are resolved now. Can you take another look?